feat: adding global.imageRegistry and global.imageTag parameters

ytimocin opened the pull request:
# Description

## Human Description

- global.imageRegistry parameter
- global.imageTag parameter
- Helm unit tests for the parameters above 
- Adding Helm unit tests to CI/CD pipeline
- Update the following commands to accept the new paramaters:
  - `rad init`
  - `rad install kubernetes`
  - `rad upgrade kubernetes`
- Adding the following functionalities to `rad init`
  - `--set` so that some parameters could be set
  - `--set-file` to initialize with custom values from a file

## Testing

You need to create a binary from this branch. Then, you can run the following commands:

- `rad install kubernetes --set global.imageRegistry=ghcr.io/username --set global.imageTag=0.45`
- You can do different combinations of the new parameters and these commands: `rad init`, `rad install kubernetes`, and `rad upgrade kubernetes`.

**Note: You should pass in `--chart deploy/Chart` because the chart that is in ghcr doesn't have the new changes.**

## AI-Generated Description

This pull request introduces support for custom container registries and image tags across all Radius Helm chart deployments, making it easier to deploy in air-gapped or private registry environments. It also adds Helm chart unit tests to the CI workflows and Makefile, and provides documentation updates for these new features and testing processes.

**Helm Chart Image Registry and Tag Support:**

* Added a new `radius.image` Helm helper template to construct fully qualified image names, supporting custom registries and tags via `global.imageRegistry` and `global.imageTag` values. All deployments and statefulsets now use this helper for their image references. [[1]](diffhunk://#diff-0ba89421395856d6246fed28806565a0d3475d08fdd81093c97091a888ef6529R44-R100) [[2]](diffhunk://#diff-6e17741710603ff5eab2c2df3885c905ebc5c657cf16232734059f645f97397fL32-R32) [[3]](diffhunk://#diff-6e17741710603ff5eab2c2df3885c905ebc5c657cf16232734059f645f97397fL41-R41) [[4]](diffhunk://#diff-fa8b3636d5cfdf8b941b89a5af4b2f3725e7ddc086f406111a08afbe14ce4267L27-R27) [[5]](diffhunk://#diff-9366e0a3212639c13de30e1ab3896ba45398986535acd618702b94c38ecab701L29-R29) [[6]](diffhunk://#diff-aa649ac3a27b3a57cd4da52d741118f2870fd721886d603f5091a90fc825e025L44-R44) [[7]](diffhunk://#diff-bb1c415d2aa24ec5f2f090395448250831b3b467289e1fa6c603fd1661453de1L122-R122) [[8]](diffhunk://#diff-92bd08c7c2302e12d2ceb2f7bf0a5bdd0016bb8e2eb2639588e994366e9ee841L122-R122) [[9]](diffhunk://#diff-0fe5610e45572cab258709e7346308d2bdc2f9c16a46d493f049de4707ea465dL35-R35)
* Updated `deploy/Chart/README.md` with detailed instructions on configuring custom registries and tags, air-gapped deployments, and certificate handling. [[1]](diffhunk://#diff-3f9a3e2ff0e91c183113a295169a98b4ec58c328613c999c82cd43e1e822d6a3L1-R149) [[2]](diffhunk://#diff-3f9a3e2ff0e91c183113a295169a98b4ec58c328613c999c82cd43e1e822d6a3L39-R176)

**Testing and CI Improvements:**

* Added Helm chart unit tests using the `helm-unittest` plugin, with sample tests and documentation in `deploy/Chart/tests/README.md`.
* Integrated Helm unit tests into CI workflows (`.github/workflows/build.yaml` and `.github/workflows/lint.yaml`) and the Makefile (`build/test.mk`). [[1]](diffhunk://#diff-d0777657fa3fd81d23aaf7273e58aee453b04e67882517900c56daeef9b3e4c1R242-R245) [[2]](diffhunk://#diff-4b122024a3a28ded65da76a2f1bface1f3a27328374438d1298d25585fe0603bR72-R75) [[3]](diffhunk://#diff-607e82dd7ce2736355a7703fb1a4a832dddb294c3683c4b51a5118218e379712L46-R46) [[4]](diffhunk://#diff-607e82dd7ce2736355a7703fb1a4a832dddb294c3683c4b51a5118218e379712R145-R151)

## Type of change

- This pull request adds or changes features of Radius and has an approved issue (issue link required).

Fixes: #issue_number

## Contributor checklist
Please verify that the PR meets the following requirements, where applicable:

<!--
This checklist uses "TaskRadio" comments to make certain options mutually exclusive.
See: https://github.com/mheap/require-checklist-action?tab=readme-ov-file#radio-groups
For details on how this works and why it's required.
-->

- An overview of proposed schema changes is included in a linked GitHub issue.
    - [ ] Yes <!-- TaskRadio schema -->
    - [x] Not applicable <!-- TaskRadio schema -->
- A design document PR is created in the [design-notes repository](https://github.com/radius-project/design-notes/), if new APIs are being introduced.
    - [ ] Yes <!-- TaskRadio design-pr -->
    - [x] Not applicable <!-- TaskRadio design-pr -->
- The design document has been reviewed and approved by Radius maintainers/approvers.
    - [ ] Yes <!-- TaskRadio design-review -->
    - [x] Not applicable <!-- TaskRadio design-review -->
- A PR for the [samples repository](https://github.com/radius-project/samples) is created, if existing samples are affected by the changes in this PR.
    - [ ] Yes <!-- TaskRadio samples-pr -->
    - [x] Not applicable <!-- TaskRadio samples-pr -->
- A PR for the [documentation repository](https://github.com/radius-project/docs) is created, if the changes in this PR affect the documentation or any user facing updates are made.
    - [ ] Yes <!-- TaskRadio docs-pr -->
    - [x] Not applicable <!-- TaskRadio docs-pr -->
- A PR for the [recipes repository](https://github.com/radius-project/recipes) is created, if existing recipes are affected by the changes in this PR.
    - [ ] Yes <!-- TaskRadio recipes-pr -->
    - [x] Not applicable <!-- TaskRadio recipes-pr -->

ytimocin committed:
feat: adding global.imageRegistry parameter

Signed-off-by: ytimocin <ytimocin@microsoft.com>

ytimocin committed:
feat: add global.imageTag parameter for Helm chart

Add global.imageTag parameter to allow setting a consistent image tag across all Radius components. This complements the existing global.imageRegistry parameter.

Priority for tag selection:
1. Component-specific tag (e.g., controller.tag)
2. global.imageTag
3. Chart AppVersion (default)

- Update all deployment templates to support global.imageTag
- Add comprehensive Helm unit tests
- Update CLI examples for install, upgrade, and init commands
- Add documentation in Chart README

Signed-off-by: ytimocin <ytimocin@microsoft.com>

ytimocin committed:
feat: simplify global.imageRegistry to use direct image names

- Update values.yaml to use simple image names (controller, ucpd, etc.)
- Modify _helpers.tpl to default to ghcr.io/radius-project namespace
- Update help messages to clarify custom registry usage
- Fix helm unit tests to match new image structure

This allows users to use custom registries more easily:
- Default: ghcr.io/radius-project/controller
- Custom: myregistry.io/controller (not myregistry.io/radius-project/controller)

Signed-off-by: ytimocin <ytimocin@microsoft.com>

ytimocin committed:
docs: Update Chart README to reflect simplified image naming

When using custom registries, images are now pulled directly from
<registry>/<image-name> format without the radius-project namespace.
Updated the air-gapped environment setup documentation to reflect
this simplified structure.

Signed-off-by: Yetkin Timocin <yetkintimocin@outlook.com>
Signed-off-by: ytimocin <ytimocin@microsoft.com>

brooke-hamilton reviewed (changes requested):
⚠️ Update: I can see that when I started reviewing this PR, the branch was on commit `af90bb8`, and then there was a force push after that. @ytimocin if there were changes that would affect the comments below let me know and I will re-test.

How would a user authenticate to a private repo that requires authentication?

I think the image tag parameter is not working. Using a local build from the PR branch, I ran this:

```shell
#!/bin/bash

set -e

k3d cluster delete
k3d cluster create -p "8081:80@loadbalancer" \
    --k3s-arg "--disable=traefik@server:*" \
    --k3s-arg "--disable=servicelb@server:*"

rad install kubernetes --set global.imageTag="0.48"
```

`rad version` returns this:

```shell
$ rad version
CLI Version Information:
RELEASE   VERSION                  BICEP     COMMIT
edge      v0.49.0-rc1-21-gaf90bb8  0.37.4    af90bb81947d4b22ae1df0ed0a8d3924d906ac88

Control Plane Information:
STATUS     VERSION
Installed  0.42.42-dev
```

`kubectl describe pod controller -n radius-system` returns this for pod image:

```shell
...
Containers:
  controller:
    Container ID:  containerd://c61f83038eafaf231134098e2d81025b87234c26661c6f903df09c1bed87a530
    Image:         ghcr.io/radius-project/controller:latest
    Image ID:      ghcr.io/radius-project/controller@sha256:32bdb60fafa4e927e8a2206f8486cc9a6704be495ecd9aaab007fad6fbed4c3c
...
```

I got the same result and no error when I used an image tag that is wrong, like this: `rad install kubernetes --set global.imageTag="banana"`

ytimocin commented:
> How would a user authenticate to a private repo that requires authentication?

Created this issue as a follow-up that I will work on next week: https://github.com/radius-project/radius/issues/10183.

ytimocin commented:
> I think the image tag parameter is not working. Using a local build from the PR branch, I ran this:

You need to pass in the chart because `rad install` uses the current chart from ghcr which doesn't have the changes.

ytimocin committed:
fix: correct image tag format in examples and remove duplicate tests

- Update examples to use actual GHCR tag format (0.48 instead of v0.48.0)
- Remove duplicate Helm unit tests from build workflow

Co-Authored-By: Yetkin Timocin <yetkin@github.com>
Signed-off-by: ytimocin <ytimocin@microsoft.com>

ytimocin commented on .github/workflows/lint.yaml:
```diff
@@ -69,6 +69,10 @@ jobs:
       - name: Run Helm linter
         run: |
           helm lint ${{ env.HELM_CHARTS_DIR }}
+      - name: Run Helm unit tests
```
Removed it from build.yaml since Helm unit tests are more of a linting/validation check for the chart templates.

brooke-hamilton commented:
> > I think the image tag parameter is not working. Using a local build from the PR branch, I ran this:
> 
> You need to pass in the chart because `rad install` uses the current chart from ghcr which doesn't have the changes.

That makes sense. I will re-test.

brooke-hamilton reviewed (changes requested):
Very minor changes needed on this - just the formatting of the version numbers on the help text.

All of these scenarios worked as expected:

```shell
rad install kubernetes \
    --chart /workspace/radius-project/radius/deploy/Chart \
    --set dashboard.enabled=false \
    --skip-contour-install

rad install kubernetes \
    --chart /workspace/radius-project/radius/deploy/Chart \
    --set global.imageTag="0.48" \
    --set de.tag=0.47 \
    --set dashboard.enabled=false \
    --skip-contour-install

rad install kubernetes \
    --chart /workspace/radius-project/radius/deploy/Chart \
    --set global.imageRegistry=ghcr.io/radius-project \
    --set global.imageTag=0.48 \
    --set de.tag=0.47 \
    --set dashboard.enabled=false \
    --skip-contour-install

rad install kubernetes \
    --chart /workspace/radius-project/radius/deploy/Chart \
    --set global.imageRegistry=ghcr.io/radius-project \
    --set global.imageTag=0.48 \
    --set de.tag=0.47 \
    --set dashboard.enabled=false \
    --skip-contour-install
```

I was not able to test the upgrade scenario because a local build sees the helm chart version as 0.42, and preflight checks fail. However, I think this is a side effect of using a local build rather than a bug.

```shell
$ rad upgrade kubernetes \
  --chart /workspace/radius-project/radius/deploy/Chart \
  --set global.imageTag=0.48

Edge build detected. Upgrading to latest stable version...
Resolved to version: 0.49.0
Current Radius version: 0.42.42-dev
Target Radius version: 0.49.0
Running pre-flight checks...
  Running Kubernetes Connectivity...
    ✓ Connected (version: v1.31.5+k3s1) with sufficient permissions
  Running Helm Connectivity...
    ✓ Helm successfully connected to cluster and found Radius release (version: 0.42.42-dev)
  Running Radius Installation...
    ✓ Radius is installed (version: 0.42.42-dev), Contour is not installed (will be installed during upgrade)
  Running Version Compatibility...
    ✗ [ERROR] Only incremental version upgrades are supported. Expected next version: 0.43.0
Error: preflight checks failed: pre-flight check 'Version Compatibility' failed: Only incremental version upgrades are supported. Expected next version: 0.43.0
```

ytimocin committed:
fix: correct image tag format in examples and remove duplicate tests

Signed-off-by: ytimocin <ytimocin@microsoft.com>

brooke-hamilton reviewed (approved):
🚀 🚢
